Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs

```diff
@@ -0,0 +1,142 @@
+        }
+
+        private bool AddPipToGraph(string fragmentName, Pip pip)
+        {
+            try
+            {
+                PipId originalPipId = pip.PipId;
+                pip.ResetPipIdForTesting();
+                bool added = false;
+                switch (pip.PipType)
+                {
+                    case PipType.Module:
+                        var modulePip = pip as ModulePip;
+                        added = m_pipGraph.AddModule(modulePip);
+                        break;
+                    case PipType.SpecFile:
+                        var specFilePip = pip as SpecFilePip;
+                        added = m_pipGraph.AddSpecFile(specFilePip);
+                        break;
+                    case PipType.Value:
+                        var valuePIp = pip as ValuePip;
+                        added = m_pipGraph.AddOutputValue(valuePIp);
+                        break;
+                    case PipType.Process:
+                        var p = pip as Process;
+                        added = m_pipGraph.AddProcess(p, default);
+                        if (p.IsService)
+                        {
+                            m_fragmentContext.AddPipIdMapping(originalPipId.Value, p.PipId.Value);
+                        }
+
+                        break;
+                    case PipType.CopyFile:
+                        var copyFile = pip as CopyFile;
+                        added = m_pipGraph.AddCopyFile(copyFile, default);
+                        break;
+                    case PipType.WriteFile:
+                        var writeFile = pip as WriteFile;
+                        added = m_pipGraph.AddWriteFile(writeFile, default);
+                        break;
+                    case PipType.SealDirectory:
+                        var sealDirectory = pip as SealDirectory;
+                        if (sealDirectory.Kind == SealDirectoryKind.Opaque || sealDirectory.Kind == SealDirectoryKind.SharedOpaque)
+                        {
+                            return true;
+                        }
+
+                        added = true;
+                        var oldDirectory = sealDirectory.Directory;
+                        sealDirectory.ResetDirectoryArtifact();
+                        var mappedDirectory = m_pipGraph.AddSealDirectory(sealDirectory, default);
+                        m_fragmentContext.AddDirectoryMapping(oldDirectory, mappedDirectory);
+                        break;
+                }
+
+                if (!added)
+                {
+                    Logger.Log.FailedToAddFragmentPipToGraph(m_loggingContext, fragmentName, pip.GetDescription(m_context));
+                    return false;
+                }
+
+                return true;
+            }
+            catch (Exception)
+            {
+                Logger.Log.FailedToAddFragmentPipToGraph(m_loggingContext, fragmentName, pip.GetDescription(m_context));
+                throw;

```

Comment:
"log and (re)throw" anti pattern.   I would remove this try/catch completely.

Reply:
how would you suggest to handle this case?  Whatever is above this is somehow swallowing exceptions.  It basically finished evaluation phase and complained that no pips matched the failure.  Feels like those exceptions should be shown somewhere.

---
In reply to: [289550605](https://github.com/microsoft/BuildXL/pull/394#discussion_r289550605) [](ancestors = 289550605)

Reply:
Yeah, I'd expect those exceptions to bubble up to the top and be generically handled there.  But I could be wrong.